### PR TITLE
ONNX model editor class with input types modification support

### PR DESCRIPTION
**Pull request by tomdol**
This PR contains a skeleton of the editor class that can be used to modify ONNX models before they are converted to `ngraph::Function` using the onnx_importer. The class operates on a parsed in-memory representation of the model which is later passed directly to the onnx_importer API.
Future extensions will be added in separate PRs (for example input shape modifications).

The class is designed the way it is to avoid the necessity of adding extra include directories to the consumers of the onnx_importer API (headers). This is why there are forward declarations for ONNX symbols in some headers. For the same reason the ModelProto object is held by a raw pointer in the editor class.

### CODE REVIEW

File: ngraph/test/onnx/onnx_editor.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+namespace
+{
+    using InputTypePred = std::function<bool(const std::shared_ptr<ngraph::Node>)>;
+
+    // A higher order factory function that produces predicates bound to a particular element type
+    InputTypePred element_type_is(const element::Type et)
+    {
+        return [et](const std::shared_ptr<ngraph::Node> input) {
+            return std::dynamic_pointer_cast<onnx_import::default_opset::Parameter>(input)
```
Comment (mbencer) [inline ngraph/test/onnx/onnx_editor.cpp:41]: I am not sure why this dynamic to `Parameter` is needed, as I remember `get_element_type` is defined in `ngraph::Node`

File: ngraph/test/onnx/onnx_editor.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    const auto graph_inputs = function->get_parameters();
+
+    const auto float_inputs_count = std::count_if(
+        std::begin(graph_inputs), std::end(graph_inputs), element_type_is(element::f32));
+
+    const auto integer_inputs_count = std::count_if(
+        std::begin(graph_inputs), std::end(graph_inputs), element_type_is(element::i64));
+
+    EXPECT_EQ(float_inputs_count, 0);
+    EXPECT_EQ(integer_inputs_count, 3);
```
Comment (mbencer) [inline ngraph/test/onnx/onnx_editor.cpp:66]: I think that it is worth to add one test which confirm that exactly type of "A" parameter was changed. I can be sth like (not tested):
```
    const auto param_a =
        std::find_if(graph_inputs.begin(), graph_inputs.end(), [](const std::shared_ptr<Parameter>& param) {
            return std::string{param->get_friendly_name()} == "A";
        });
EXPECT_EQ(param_a->get_element_type(), element::i64);
```

**Review by mbencer — APPROVED** [review]

File: ngraph/frontend/onnx_import/src/editor/editor.cpp [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    void modify_input_type(ValueInfoProto& onnx_input, element::Type_t elem_type)
+    {
+        const std::string malformed_input_error_msg =
+            "The input is malformed, cannot change the data type. Input name: " + onnx_input.name();
+
+        if (!onnx_input.has_type())
+        {
+            throw ngraph_error(malformed_input_error_msg);
```
Comment (mateusztabaka) [inline ngraph/frontend/onnx_import/src/editor/editor.cpp:64]: I think on the negative paths, we can have a specific error message (e.g. here "Input is malformed - it has no type" and similar in line 71) so the error message is not created every time  modify_input_type is called, but only on the negative path (which should be rare).

**Review by mateusztabaka — APPROVED** [review]

File: ngraph/frontend/onnx_import/src/editor/editor.cpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                auto* tensor_type = type_proto->mutable_tensor_type();
+                if (NG_2_ONNX_TYPES.count(elem_type) == 0)
+                {
+                    throw ngraph_error("The input type for input '" + onnx_input.name() +
+                                       "' cannot be set to: " +
+                                       element::Type(elem_type).get_type_name() +
+                                       ". This type is not allowed in ONNX.");
+                }
+                else
+                {
```
[the remark this answers is not in the document]
Comment (tomdol) [reply ngraph/frontend/onnx_import/src/editor/editor.cpp]: I've implemented the suggestions above but in this particular place I think it's more readable if the else branch is left as-is.

File: ngraph/frontend/onnx_import/src/editor/editor.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            }
+        }
+
+        return nullptr;
+    }
+
+    void modify_input_type(ValueInfoProto& onnx_input, element::Type_t elem_type)
+    {
+        const std::string malformed_input_error_msg =
+            "The input is malformed, cannot change the data type. Input name: " + onnx_input.name();
```
Comment (mateusztabaka) [inline ngraph/frontend/onnx_import/src/editor/editor.cpp:60]: this variable can be removed now

**Review by tsocha — APPROVED** [review]

**Review by postrational — COMMENTED** [review]
Let's try to apply the PImpl idiom and perhaps we'll be able to remove `new`, `delete` and the return value parameter in `parse_from_` functions.

File: ngraph/frontend/onnx_import/include/onnx_import/editor/editor.hpp [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            ONNX_NAMESPACE::ModelProto& model() const;
+
+            /// \brief Returns the path to the original model file
+            const std::string& model_path() const;
+
+        private:
+            const std::string m_model_path;
+
+            class Impl;
+            std::unique_ptr<Impl, void (*)(Impl*)> m_pImpl;
```
Comment (postrational) [inline ngraph/frontend/onnx_import/include/onnx_import/editor/editor.hpp:76]:
```suggestion
            std::unique_ptr<Impl, void (*)(Impl*)> m_pimpl;
```

**Review by postrational — APPROVED** [review]
Looks good!

File: ngraph/frontend/onnx_import/include/onnx_import/editor/editor.hpp [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+{
+    namespace onnx_import
+    {
+        /// \brief A class representing a set of utilities allowing modification of an ONNX model
+        ///
+        /// \note This class can be used to modify an ONNX model before it gets translated to
+        ///       an ngraph::Function by the import_onnx_model function. It lets you modify the
+        ///       model's input types and shapes, extract a subgraph and more. An instance of this
+        ///       class can be passed directly to the onnx_importer API.
+        class ONNX_IMPORTER_API ONNXModelEditor final
```
Comment (ilyachur) [inline ngraph/frontend/onnx_import/include/onnx_import/editor/editor.hpp:43]: I have a question 😄 
Why do we need this class and who will use it in the Inference Engine?
As I know we use ONNX reader in order to read the ONNX model, will the reader use it? Or not?

Comment (tomdol) [reply ngraph/frontend/onnx_import/include/onnx_import/editor/editor.hpp:43]: This class is one of the steps to acchieve MO-parity. It's supposed to be used to adjust the model before it's imported in many ways. This PR only add support for input type changing, but the upcoming work will deliver reshapeability, model cutting and more.

File: ngraph/frontend/onnx_import/include/onnx_import/utils/parser.hpp [hunk 8]
```diff
@@ -0,0 +1,45 @@
+//*****************************************************************************
+// Copyright 2017-2020 Intel Corporation
```
Comment (mitruska) [inline ngraph/frontend/onnx_import/include/onnx_import/utils/parser.hpp:2]: Just friendly reminder about the new year, same for other files:)
```suggestion
// Copyright 2017-2021 Intel Corporation
```

**Review by mitruska — APPROVED** [review]

File: ngraph/frontend/onnx_import/include/onnx_import/utils/parser.hpp (hunk 8 above)
Comment (tomdol) [reply ngraph/frontend/onnx_import/include/onnx_import/utils/parser.hpp:2]: I've created a PR with batch-update of copyright information https://github.com/openvinotoolkit/openvino/pull/3832

**Pull request merged**